### ISSUE TITLE
autoimage seems off center

### ISSUE THREAD

**Issue by Hendejac** [Amber-MD/cpptraj#860]
I am trying to rewrap the water in my system because my protein partially leaves the system boundary. The dcd trajectory files come from CHARMM version 36a6 and a truncated octahedron periodic boundary condition was used. I used the following as input for cpptraj (V4.12.1).

`parm protein.psf `
`trajin protein.dcd`
`autoimage`
`trajout output.dcd`
`go`
`exit`

Here is a snapshot from the output.dcd file so you can see the issue I am having. It appears that the truncated octahedron is offset a little bit and you are seeing corners of what might be the neighboring images.

<img width="374" alt="Screen Shot 2020-11-09 at 2 48 12 PM" src="https://user-images.githubusercontent.com/45015829/98595490-0d6a8e00-22a4-11eb-8193-8e09e9e726f6.png">

I have tried changing the anchor selections, as well as trying origin and firstatom in the autoimage command, but each time I get the same issue to some degree. Any suggestions on how to fix this problem would be greatly appreciated. 

I want the water rewrapped because I am interested in looking at the water-mediated hydrogen bonds in and around this particular protein. If it is true that the resulting dcd file is being reimaged (rewrapped) slightly off-center and that the waters immediately around the protein are correctly placed then the strange behavior seen around the edges of the rewrapped system would not lead to inaccuracy in my analysis. This is what looks to be the case, but I want to make sure that the water placements are not in general incorrect. 

Thank you greatly.

**Comment by drroe** [comment]
Any chance you could provide me with the PSF and a few frames of the trajectory so I can try to reproduce? Thanks.

**Comment by Hendejac** [comment]
Yes. I have added some files to a google drive, I have attached a link below. The files in the google drive include the pdb and psf file of the system, a dcd from the beginning of the simulation (before the protein leaves the boundary), and a dcd from the end of the simulation (after parts of the protein have crossed the boundary). These simulations are from replica exchange so they are not continuous trajectories.

https://drive.google.com/drive/folders/15N6AtDF3knBansKza53khAlFkiL15866?usp=sharing 

Thank you.

**Comment by drroe** [comment]
Sorry for the delay on this. I'm currently on parental leave so my response time is down. Will try to look at it today and tomorrow.

**Comment by drroe** [comment]
Thanks for the files. One question - are the trajectories you provided directly from CHARMM `DYNA`, or have they been processed by something else (CHARMM, cpptraj, etc) in the meantime?

**Comment by Hendejac** [comment]
They are directly from CHARMM Dyna

**Comment by drroe** [comment]
One last question - was this from constant pressure dynamics? I'm still testing but I think the issue is how the shape matrix is translated to an Alpha Beta Gamma reference frame.

**Comment by Hendejac** [comment]
Yes, this is from constant pressure dynamics.

**Comment by drroe** [comment]
OK I have good news and bad news. The good news is I think I know what's wrong. Internally, cpptraj uses a fixed x-y-z-alpha-beta-gamma reference for unit cells. This is fine for almost all simulations. However, charmm (and maybe other codes, I'm not 100% sure) uses a symmetric shape matrix. In most cases the two are equivalent, but for constant pressure runs the charmm integrator works in such a way that the unit cell basis can rotate. When cpptraj converts the shape matrix to the x-y-z-a-b-g frame, any such rotation is lost, and imaging becomes incorrect. When I use the shape matrix directly, things seem much better.

The bad news is that this is a pretty major change under the hood, so it will take a while to implement, and there is a chance it could break pytraj (pinging @hainm so he's aware). So I'll be working on this but it may be that you need to use code from a development branch before I can get a fix merged in. Sorry for the delay.

I'm working in this branch: https://github.com/drroe/cpptraj/tree/fixshapematrix

You're welcome to try it out but be aware it's very very alpha and so I can't guarantee results yet.

**Comment by drroe** [comment]
I'm still working on this. About 90% done; right now the delay is in all the testing. Imaged distance calculations are central to many actions, so I am making sure that I don't break anything and also don't kill performance. Right now it seems worst case scenario is about a 1% speed hit, and for many cases there may be no noticeable loss in performance. I hope to have the changes finished by next week.

**Comment by Hendejac** [comment]
Awesome! Thank you!

**Comment by drroe** [comment]
I'm going to try to merge this today (or at least start the merge request). As I thought it was a major change (in fact, this is the first major version increase in a long time). However, I think it's very worth it in the long run. Performance does not suffer too badly, and actually is boosted for some actions. And now rotations (from e.g. rms, principal, etc) will no longer break imaging. I also managed to fix a few bugs while I was in the code.

@hainm the bad news is that (as you might expect) this breaks pytraj. I've started to try and fix this (https://github.com/drroe/pytraj/tree/fixshapematrix) but it may be easier for you to do it. I'll ping you again once the merge request is up.
